### ISSUE TITLE
Support Unpartitioned Analytic Invocation

### ISSUE THREAD

**Issue by noahboerger** [sdmx-twg/vtl#520]
UM|189|partitionClause
|:--|--:|--:
***

### Issue Description

The partitionClause is defined in the reference manual 2.1 (p. 189) as `partitionClause ::= partition by identifier { , identifier }*` with the additional description of the default being _If the clause is omitted, then the Data Set is partitioned by the Identifier Components that are not specified in the orderClause_. 

Therefore it is currently not possible to perform an analytic invocation without partitioning.

### Proposed Solution

Support partitionClause without specifing any identifier or set the default when omitting the partitioningClause to this behaviour, to allow analytic invocations on unpartitioned datasets.

### Example

When `ds1` is

| id_1       | id_2       | id_3      | val     | 
|------------|------------|---------|---------|
| IDENTIFIER | IDENTIFIER | IDENTIFIER | MEASURE| 
| INTEGER    | INTEGER    | INTEGER |  INTEGER|
| 1          | 1          | 1       |  1 |
| 1          | 1          | 2       |  1 |
| 1          | 2          | 1       |  1 |
| 2          | 1          | 1       |  1 |
| 2          | 2          | 1       |  1 |
| 2          | 2          | 2       |  1 |
| 3          | 1          | 1       |  1 |
| 3          | 2          | 1       |  1 |
| 3          | 2          | 2       |  1 |

The result of `ds2 := ds1[calc rank_col := rank(over(order by id_1))];` or `ds2 := ds1[calc rank_col := rank(over(partition by order by id_1))];` should be out of my point of view

| id_1       | id_2       | id_3      | val     | rank_col | 
|------------|------------|---------|---------|---------|
| IDENTIFIER | IDENTIFIER | IDENTIFIER | MEASURE | MEASURE | 
| INTEGER    | INTEGER    | INTEGER |  INTEGER | INTEGER |
| 1          | 1          | 1       |  1 | 1 |
| 1          | 1          | 2       |  1 | 1 |
| 1          | 2          | 1       |  1 | 1 |
| 2          | 1          | 1       |  1 |  4 |
| 2          | 2          | 1       |  1 |  4 |
| 2          | 2          | 2       |  1 | 4 |
| 3          | 1          | 1       |  1 | 7 |
| 3          | 2          | 1       |  1 | 7 |
| 3          | 2          | 2       |  1 | 7 |

but currently is (because of the default partitioning / not supported gramatic to have a partitionClause without specifying any identifier)

| id_1       | id_2       | id_3      | val     | rank_col | 
|------------|------------|---------|---------|---------|
| IDENTIFIER | IDENTIFIER | IDENTIFIER | MEASURE | MEASURE | 
| INTEGER    | INTEGER    | INTEGER |  INTEGER | INTEGER |
| 1          | 1          | 1       |  1 | 1 |
| 1          | 1          | 2       |  1 | 1 |
| 1          | 2          | 1       |  1 | 1 |
| 2          | 1          | 1       |  1 |  2 |
| 2          | 2          | 1       |  1 |  2 |
| 2          | 2          | 2       |  1 | 1 |
| 3          | 1          | 1       |  1 | 3 |
| 3          | 2          | 1       |  1 | 3 |
| 3          | 2          | 2       |  1 | 2 |

**Comment by antonio-olleros** [comment]
I agree, this seems like a limitation of the language, because it is not possible to get a result that may need to be calculated

**Comment by linardian** [comment]
Hi @noahboerger. First of all (sorry, but) I do not understand the results you are expecting, so if you can please elaborate a little bit more on this I would be very grateful. Secondly, the behaviour of the operator is exactly the one adopted by RDBMS (I made a test with ORACLE and DB2). Thank you so much.

**Comment by noahboerger** [comment]
Hi @linardian,

Thank you for your response.

To clarify my request, I aim to perform an analytic operation equivalent to the following SQL query:

`SELECT  
    id_1,  
    id_2,  
    id_3,  
    val,  
    RANK() OVER (ORDER BY id_1) AS rank  
FROM  
    ds1`

However, using the current VTL statement:

`ds1[calc rank_col := rank(over(order by id_1))]; `

instead, it translates to a SQL query with **default partitioning based on the identifiers** not specified in the ORDER BY clause, equal to:

 `SELECT  
    id_1,  
    id_2,  
    id_3,  
    val,  
    RANK() OVER (PARTITION BY id_2, id_3 ORDER BY id_1) AS rank_within_partition  
FROM  
    ds1` 

This default behavior enforces partitioning by id_2 and id_3, which restricts the rank calculation to within each partition rather than across the entire dataset.

Additionally, attempting to disable partitioning by using an empty PARTITION BY clause:

`ds1[calc rank_col := rank(over(partition by order by id_1))]; `

is currently not supported, as the grammar does not allow a partition by without specifying any identifiers.

Am I missing a way to execute a VTL equivalent calculation as the SQL statement?

**Proposed Enhancement:**

To achieve the desired analytic operation without default partitioning, it would be beneficial to allow an PARTITION BY clause in VTL without specifying any identifiers. This enhancement would enable users to perform analytic functions across the entire dataset without being constrained by implicit partitioning based on unspecified identifiers.

**Comment by linardian** [comment]
Ok, now I understand.
Well, since:

SELECT   id_1,   id_2,   id_3,   val,   RANK() OVER (ORDER BY id_1) AS rank   FROM   ds1
 
is equivalent to:

select id1, id2, id3, val, RANK() OVER (partition by id4 ORDER BY id1) AS rank
from (SELECT id1, id2, id3, val as id4, val FROM ds1)

you can easily get the same result adding a new id: 

ds1[calc identifier id_4 := val];

and then calculating the rank().
Is this correct ?

**Comment by noahboerger** [comment]
Yes, it is possible to create an identifier that is same for all rows and use it in the partitionClause. Thank you for the workaround. However, I find this approach neither intuitive nor performance-efficient. I would still prefer to directly allow unpartitioned analytical invocations.

E.g. could it be a solution to change the grammar in 'partitionClause ::= partition by identifier { , identifier }*' to make also the first identifier optional? What is your opinion on this topic?

**Comment by linardian** [comment]
Ok. I have added this issue in the topics of the next meeting in Frankfurt
